File: packages/python/src/agnes_ai/_http.py
from __future__ import annotations

import json
import time
from collections.abc import Iterator, Mapping
from typing import Any

import httpx

from .config import AgnesConfig
from .constants import RETRYABLE_STATUS_CODES
from .errors import (
    AgnesAPIAuthenticationError,
    AgnesAPIBadRequestError,
    AgnesAPIError,
    AgnesAPIRateLimitError,
    AgnesAPIServerError,
    AgnesAPITimeoutError,
)


class AgnesHTTPClient:
    def __init__(
        self,
        config: AgnesConfig,
        *,
        transport: httpx.BaseTransport | None = None,
    ) -> None:
        self._config = config
        self._client = httpx.Client(
            base_url=config.base_url,
            timeout=config.timeout,
            transport=transport,
        )

    def close(self) -> None:
        self._client.close()

    def request(
        self,
        method: str,
        endpoint: str,
        *,
        json_body: Mapping[str, Any] | None = None,
        params: Mapping[str, Any] | None = None,
    ) -> dict[str, Any]:
        response = self._send_with_retries(
            method,
            endpoint,
            json_body=json_body,
            params=params,
        )
        self._raise_for_status(response, endpoint)
        return self._parse_json_response(response, endpoint)

    def stream(
        self,
        method: str,
        endpoint: str,
        *,
        json_body: Mapping[str, Any] | None = None,
        params: Mapping[str, Any] | None = None,
    ) -> Iterator[str]:
        try:
            with self._client.stream(
                method,
                endpoint,
                headers=self._headers(),
                json=json_body,
                params=params,
            ) as response:
                self._raise_for_status(response, endpoint)
                for chunk in response.iter_text():
                    if chunk:
                        yield chunk
        except httpx.TimeoutException as exc:
            raise AgnesAPITimeoutError(
                "Agnes API request timed out.",
                endpoint=endpoint,
            ) from exc

    def _send_with_retries(
        self,
        method: str,
        endpoint: str,
        *,
        json_body: Mapping[str, Any] | None,
        params: Mapping[str, Any] | None,
    ) -> httpx.Response:
        last_response: httpx.Response | None = None

        for attempt in range(self._config.max_retries + 1):
            try:
                response = self._client.request(
                    method,
                    endpoint,
                    headers=self._headers(),
                    json=json_body,
                    params=params,
                )
            except httpx.TimeoutException as exc:
                raise AgnesAPITimeoutError(
                    "Agnes API request timed out.",
                    endpoint=endpoint,
                ) from exc

            if response.status_code not in RETRYABLE_STATUS_CODES:
                return response

            last_response = response
            if attempt < self._config.max_retries:
                time.sleep(self._config.retry_backoff * (2**attempt))

        assert last_response is not None
        return last_response

    def _headers(self) -> dict[str, str]:
        return {
            **dict(self._config.default_headers),
            "Authorization": f"Bearer {self._config.api_key}",
            "Content-Type": "application/json",
        }

    @staticmethod
    def _parse_json_response(response: httpx.Response, endpoint: str) -> dict[str, Any]:
        try:
            data = response.json()
        except json.JSONDecodeError as exc:
            raise AgnesAPIError(
                "Agnes API returned a non-JSON response.",
                status_code=response.status_code,
                endpoint=endpoint,
                request_id=response.headers.get("x-request-id"),
            ) from exc

        if isinstance(data, dict):
            return data

        raise AgnesAPIError(
            "Agnes API returned an unexpected JSON response.",
            status_code=response.status_code,
            endpoint=endpoint,
            request_id=response.headers.get("x-request-id"),
        )

    @staticmethod
    def _raise_for_status(response: httpx.Response, endpoint: str) -> None:
        if response.status_code < 400:
            return

        message = _safe_error_message(response)
        request_id = response.headers.get("x-request-id")
        kwargs = {
            "status_code": response.status_code,
            "endpoint": endpoint,
            "request_id": request_id,
        }

        if response.status_code == 401:
            raise AgnesAPIAuthenticationError(message, **kwargs)
        if response.status_code == 400:
            raise AgnesAPIBadRequestError(message, **kwargs)
        if response.status_code == 429:
            raise AgnesAPIRateLimitError(message, **kwargs)
        if response.status_code >= 500:
            raise AgnesAPIServerError(message, **kwargs)
        raise AgnesAPIError(message, **kwargs)


def _safe_error_message(response: httpx.Response) -> str:
    fallback = f"Agnes API request failed with status {response.status_code}."
    try:
        data = response.json()
    except json.JSONDecodeError:
        return fallback

    if not isinstance(data, Mapping):
        return fallback

    error = data.get("error")
    if isinstance(error, Mapping):
        message = error.get("message")
        if isinstance(message, str) and message:
            return message

    message = data.get("message")
    if isinstance(message, str) and message:
        return message

    return fallback

File: packages/javascript/src/http.ts
import { RETRYABLE_STATUS_CODES } from "./constants.js";
import type { AgnesRequestOptions, ResolvedAgnesClientConfig } from "./config.js";
import {
  AgnesAPIAbortError,
  AgnesAPIAuthenticationError,
  AgnesAPIBadRequestError,
  AgnesAPIError,
  AgnesAPIRateLimitError,
  AgnesAPIServerError,
  AgnesAPITimeoutError,
} from "./errors.js";

export type JsonObject = Record<string, unknown>;

export interface RequestOptions extends AgnesRequestOptions {
  body?: JsonObject;
  params?: Record<string, string | number | boolean | undefined>;
}

export class AgnesHTTPClient {
  constructor(private readonly config: ResolvedAgnesClientConfig) {}

  async request(method: string, endpoint: string, options: RequestOptions = {}): Promise<JsonObject> {
    for (let attempt = 0; attempt <= this.config.maxRetries; attempt += 1) {
      const lifecycle = createAbortLifecycle(options.signal, this.config.timeout);
      let shouldRetry = false;

      try {
        const response = await this.send(method, endpoint, options, lifecycle.signal);
        shouldRetry = RETRYABLE_STATUS_CODES.has(response.status) && attempt < this.config.maxRetries;
        if (shouldRetry) {
          await cancelBody(response);
        } else {
          await raiseForStatus(response, endpoint, this.config.apiKey);
          return await parseJsonObject(response, endpoint);
        }
      } catch (error) {
        throw lifecycle.translate(error, endpoint);
      } finally {
        lifecycle.cleanup();
      }

      if (shouldRetry) {
        await sleep(this.config.retryBackoff * 2 ** attempt, options.signal, endpoint);
      }
    }

    throw new AgnesAPIError("Agnes API request failed without a response.", { endpoint });
  }

  async *stream(
    method: string,
    endpoint: string,
    options: RequestOptions = {},
  ): AsyncIterable<string> {
    const lifecycle = createAbortLifecycle(options.signal, this.config.timeout);
    try {
      const response = await this.send(method, endpoint, options, lifecycle.signal);
      await raiseForStatus(response, endpoint, this.config.apiKey);

      const body = response.body;
      if (!body) {
        return;
      }

      const decoder = new TextDecoder();
      for await (const chunk of body as unknown as AsyncIterable<Uint8Array>) {
        const text = decoder.decode(chunk, { stream: true });
        if (text) yield text;
      }
      const tail = decoder.decode();
      if (tail) yield tail;
    } catch (error) {
      throw lifecycle.translate(error, endpoint);
    } finally {
      lifecycle.cleanup();
    }
  }

  private send(
    method: string,
    endpoint: string,
    options: RequestOptions,
    signal: AbortSignal,
  ): Promise<Response> {
    return this.config.fetch(buildUrl(this.config.baseUrl, endpoint, options.params), {
      method,
      headers: this.headers(),
      body: options.body === undefined ? undefined : JSON.stringify(options.body),
      signal,
    });
  }

  private headers(): Record<string, string> {
    return {
      ...this.config.defaultHeaders,
      Authorization: `Bearer ${this.config.apiKey}`,
      "Content-Type": "application/json",
    };
  }
}

async function raiseForStatus(
  response: Response,
  endpoint: string,
  apiKey?: string,
): Promise<void> {
  if (response.status < 400) {
    return;
  }

  const message = await safeErrorMessage(response, { apiKey });
  const options = {
    endpoint,
    requestId: response.headers.get("x-request-id") ?? undefined,
    statusCode: response.status,
  };

  if (response.status === 401) {
    throw new AgnesAPIAuthenticationError(message, options);
  }
  if (response.status === 400) {
    throw new AgnesAPIBadRequestError(message, options);
  }
  if (response.status === 429) {
    throw new AgnesAPIRateLimitError(message, options);
  }
  if (response.status >= 500) {
    throw new AgnesAPIServerError(message, options);
  }
  throw new AgnesAPIError(message, options);
}

async function parseJsonObject(response: Response, endpoint: string): Promise<JsonObject> {
  const data = (await response.json()) as unknown;
  if (isJsonObject(data)) {
    return data;
  }

  throw new AgnesAPIError("Agnes API returned an unexpected JSON response.", {
    endpoint,
    statusCode: response.status,
  });
}

async function safeErrorMessage(
  response: Response,
  options: { apiKey?: string } = {},
): Promise<string> {
  const fallback = `Agnes API request failed with status ${response.status}.`;
  let data: unknown;

  try {
    data = await response.clone().json();
  } catch {
    return fallback;
  }

  if (!isJsonObject(data)) {
    return fallback;
  }

  const error = data.error;
  if (isJsonObject(error) && typeof error.message === "string" && error.message.length > 0) {
    return redactSecret(error.message, options.apiKey);
  }

  if (typeof data.message === "string" && data.message.length > 0) {
    return redactSecret(data.message, options.apiKey);
  }

  return fallback;
}

function redactSecret(message: string, apiKey?: string): string {
  let redacted = apiKey ? message.replaceAll(apiKey, "[REDACTED]") : message;
  redacted = redacted.replace(/\bBearer\s+[^\s,;]+/gi, "Bearer [REDACTED]");
  return redacted;
}

function buildUrl(
  baseUrl: string,
  endpoint: string,
  params?: Record<string, string | number | boolean | undefined>,
): string {
  const url = new URL(endpoint, `${baseUrl}/`);
  for (const [key, value] of Object.entries(params ?? {})) {
    if (value !== undefined) {
      url.searchParams.set(key, String(value));
    }
  }
  return url.toString();
}

function isJsonObject(data: unknown): data is JsonObject {
  return typeof data === "object" && data !== null && !Array.isArray(data);
}

interface AbortLifecycle {
  signal: AbortSignal;
  cleanup(): void;
  translate(error: unknown, endpoint: string): unknown;
}

function createAbortLifecycle(callerSignal: AbortSignal | undefined, timeout: number): AbortLifecycle {
  const controller = new AbortController();
  let abortKind: "caller" | "timeout" | undefined;

  const abort = (kind: "caller" | "timeout") => {
    if (abortKind) return;
    abortKind = kind;
    controller.abort();
  };
  const onCallerAbort = () => abort("caller");

  if (callerSignal?.aborted) {
    abort("caller");
  } else {
    callerSignal?.addEventListener("abort", onCallerAbort, { once: true });
  }
  const timeoutId = setTimeout(() => abort("timeout"), timeout);

  return {
    signal: controller.signal,
    cleanup() {
      clearTimeout(timeoutId);
      callerSignal?.removeEventListener("abort", onCallerAbort);
    },
    translate(error, endpoint) {
      if (error instanceof AgnesAPIAbortError || error instanceof AgnesAPITimeoutError) {
        return error;
      }
      if (abortKind === "caller") {
        return new AgnesAPIAbortError("Agnes API request was cancelled by the caller.", { endpoint });
      }
      if (abortKind === "timeout") {
        return new AgnesAPITimeoutError("Agnes API request timed out.", { endpoint });
      }
      return error;
    },
  };
}

async function cancelBody(response: Response): Promise<void> {
  try {
    await response.body?.cancel();
  } catch {
    // A retry is already replacing this response; cancellation is best effort.
  }
}

function sleep(ms: number, signal: AbortSignal | undefined, endpoint: string): Promise<void> {
  if (signal?.aborted) {
    return Promise.reject(new AgnesAPIAbortError("Agnes API request was cancelled by the caller.", {
      endpoint,
    }));
  }

  return new Promise((resolve, reject) => {
    const timeoutId = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(timeoutId);
      reject(new AgnesAPIAbortError("Agnes API request was cancelled by the caller.", { endpoint }));
    };
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}
